sat/metrics/metrics.go

```diff
@@ -0,0 +1,124 @@
+package metrics
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/metric/global"
+	"go.opentelemetry.io/otel/metric/instrument"
+	"go.opentelemetry.io/otel/metric/instrument/syncint64"
+	"go.opentelemetry.io/otel/metric/unit"
+	controller "go.opentelemetry.io/otel/sdk/metric/controller/basic"
+	processor "go.opentelemetry.io/otel/sdk/metric/processor/basic"
+	"go.opentelemetry.io/otel/sdk/metric/selector/simple"
+
+	"github.com/suborbital/sat/sat/options"
+)
+
+type Metrics struct {
+	FunctionExecutions       syncint64.Counter
+	FailedFunctionExecutions syncint64.Counter
+	FunctionTime             syncint64.Histogram
+}
+
+type Timer struct {
+	start time.Time
+}
+
+func (t Timer) Observe() int64 {
+	return time.Now().Sub(t.start).Milliseconds()
+}
+
+// NewTimer returns a Timer with the current time stored in it.
+func NewTimer() Timer {
+	return Timer{start: time.Now()}
+}
+
+// SetupMetricsProvider takes in an options.MetricsConfig struct and a logger to put together the structure of getting
+// measured data onto the collector. It does not set up the actual meters that we need to use to actually measure
+// anything, that's the job of ConfigureMetrics.
+//
+// This structure is configured to be in the global scope, and that's where all other meters will send their data to be
+// picked up and sent off to the collector at specified intervals.
+func SetupMetricsProvider(config options.MetricsConfig) error {
+	exporter, err := otlpmetricgrpc.New(
+		context.TODO(),
+		otlpmetricgrpc.WithTimeout(5*time.Second),
+		otlpmetricgrpc.WithRetry(otlpmetricgrpc.RetryConfig{
+			Enabled:         true,
+			InitialInterval: 2 * time.Second,
+			MaxInterval:     10 * time.Second,
+			MaxElapsedTime:  30 * time.Second,
+		}),
+		otlpmetricgrpc.WithEndpoint(config.Endpoint),
+		otlpmetricgrpc.WithInsecure(),
+	)
+	if err != nil {
+		return errors.Wrap(err, "otlpmetricgrpc.New")
+	}
+
+	cont := controller.New(
+		processor.NewFactory(
+			simple.NewWithInexpensiveDistribution(),
+			exporter,
+		),
+		controller.WithExporter(exporter),
+		controller.WithCollectPeriod(3*time.Second),
+	)
+
+	if err := cont.Start(context.Background()); err != nil {
+		return errors.Wrap(err, "metric controller Start")
+	}
+
+	global.SetMeterProvider(cont)
+
+	return nil
+}
+
+// ConfigureMetrics returns a struct with the meters that we want to measure in sat. It assumes that the global meter
+// has already been set up (see SetupMetricsProvider). Shipping the measured values is the task of the provider, so
+// from a usage point of view, nothing else is needed.
+func ConfigureMetrics() (Metrics, error) {
+	m := global.Meter(
+		"sat",
+		metric.WithInstrumentationVersion("1.0"),
+	)
+
+	si64 := m.SyncInt64()
+
+	functionExecutions, err := si64.Counter(
```

Comment:
Any reason not to use the async counter

Reply:
I tend to favor async implementations on the hot path for supporting functions wherever possible. 

One area where this becomes especially important is logging. When things start going bad logging tends to increase, if the bad things happens to be IO contention you end up slowing everything else down which can cause even more issues which increases logging, which increases contention which causes more issues which increases logging...